### PR TITLE
Throw http response content when 403

### PR DESCRIPTION
**Pull request by terencefan**
Throw http response content when 403

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRForbiddenException.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Azure.SignalR.Common;
+
+[Serializable]
+public class AzureSignalRForbiddenException : AzureSignalRException
+{
+    public const string IngressDenied = "Nginx denied the access, please check your Networking settings.";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRForbiddenException.cs]: Ingress is also one component of Azure SignalR service from customer's perspective of view, let's not expose unnecessory terms to our customers.

Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRForbiddenException.cs]: Ingress is also one component of Azure SignalR service from customer's perspective of view, let's not expose unnecessory terms to our customers.

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs [hunk 2]
```diff
@@ -195,8 +196,9 @@ private static async Task ThrowExceptionOnResponseFailureAsync(HttpResponseMessa
         var requestUri = response.RequestMessage?.RequestUri?.ToString();
         throw response.StatusCode switch
         {
-            HttpStatusCode.BadRequest => new AzureSignalRInvalidArgumentException(requestUri, innerException, detail),
+            HttpStatusCode.BadRequest => new AzureSignalRInvalidArgumentException(requestUri, innerException, content),
             HttpStatusCode.Unauthorized => new AzureSignalRUnauthorizedException(requestUri, innerException),
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:200]: add more into this UnauthorizedException is enough, for example, http status code into it?

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:200]: I am not sure what is the value added here to add kinds of exceptions? Do we expect the customer to catch these exceptions? Or it is just for printing different error messages? If the latter, I feel reuse AzureSignalRException is enough

File: src/Microsoft.Azure.SignalR.Common/Constants.cs [hunk 3]
```diff
@@ -17,6 +17,10 @@ internal static class Constants
 
         public const string AsrsIsDiagnosticClient = "Asrs-Is-Diagnostic-Client";
 
+        public const string IngressDenied = "ASRS denied the access, please check your Networking settings.";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Constants.cs]:
```suggestion
        public const string IngressDenied = "..., please check your Networking settings.";
```

File: src/Microsoft.Azure.SignalR.Common/Constants.cs [hunk 4]
```diff
@@ -17,6 +17,10 @@ internal static class Constants
...
+
+        public const string RuntimeDenied = "ASRS denied the access, please check your Access control (IAM) settings.";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Constants.cs]:
```suggestion
        public const string RuntimeDenied = "SignalR service denied the access, please check your service Access Control (IAM) settings.";
```

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs [hunk 5]
```diff
@@ -24,8 +25,8 @@ public AzureSignalRAccessTokenNotAuthorizedException(string message) : base(mess
         /// <summary>
         /// Initializes a new instance of the <see cref="AzureSignalRAccessTokenNotAuthorizedException"/> class.
         /// </summary>
-        public AzureSignalRAccessTokenNotAuthorizedException(string credentialName, Exception innerException) :
-            base(credentialName + Postfix, innerException)
+        public AzureSignalRAccessTokenNotAuthorizedException(TokenCredential credential, Exception innerException) :
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs]: breaking change

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs [hunk 6]
```diff
@@ -2,24 +2,40 @@
...
-        }
+            HttpStatusCode.Forbidden => GetForbiddenReason(content),
+            _ => ErrorMessage,
+        };
+        return string.IsNullOrEmpty(requestUri) ? reason : $"{reason} Request Uri: {requestUri}";
+    }
+
+    private static string GetForbiddenReason(string content)
+    {
+        return content.Contains("nginx") ? Constants.IngressDenied : Constants.RuntimeDenied;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs]: Have the constants into the exception sounds enough?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs [hunk 7]
```diff
@@ -2,28 +2,21 @@
...
-        private const string ErrorMessage = "Authorization failed. If you were using AccessKey, please check connection string and see if the AccessKey is correct. If you were using Azure Active Directory, please note that the role assignments will take up to 30 minutes to take effect if it was added recently.";
+namespace Microsoft.Azure.SignalR.Common;
 
-        public AzureSignalRUnauthorizedException(string requestUri, Exception innerException) : base(string.IsNullOrEmpty(requestUri) ? ErrorMessage : $"{ErrorMessage} Request Uri: {requestUri}", innerException)
-        {
-        }
+[Serializable]
+public class AzureSignalRUnauthorizedException : AzureSignalRException
+{
+    private const string ErrorMessage = "API failed with status code 401.";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs]: what does API mean? simply `401 Unauthroized`?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs (hunk 6 above)
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs]: I  prefer not depend on current implementation detail

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs (hunk 7 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs]: revert the change

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs [hunk 8]
```diff
@@ -22,11 +23,22 @@ public AzureSignalRAccessTokenNotAuthorizedException(string message) : base(mess
         }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="AzureSignalRAccessTokenNotAuthorizedException"/> class.
+        /// Obsolete, <see cref="AzureSignalRAccessTokenNotAuthorizedException(TokenCredential, Exception)"/>.
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:26]: then the other ctor shouldn't be internal?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs [hunk 9]
```diff
@@ -22,11 +23,22 @@ public AzureSignalRAccessTokenNotAuthorizedException(string message) : base(mess
...
         public AzureSignalRAccessTokenNotAuthorizedException(string credentialName, Exception innerException) :
             base(credentialName + Postfix, innerException)
         {
         }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureSignalRAccessTokenNotAuthorizedException"/> class.
+        /// </summary>
+        internal AzureSignalRAccessTokenNotAuthorizedException(TokenCredential credential, Exception innerException) :
+            base(credential.GetType().Name + Postfix, innerException)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:40]: why changing to `TokenCredential` if only Name is used?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs [hunk 10]
```diff
@@ -2,24 +2,35 @@
...
+        var reason = statusCode switch
         {
-        }
+            HttpStatusCode.Forbidden => GetForbiddenReason(content),
+            _ => ErrorMessage,
+        };
+        return string.IsNullOrEmpty(requestUri) ? reason : $"{reason} Request Uri: {requestUri}";
     }
+
+    private static string GetForbiddenReason(string content) => content;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRRuntimeException.cs]: `Response detail: {content}`?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs (hunk 8 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:26]: The customer shouldn't construct this exception by himself.

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs (hunk 9 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRAccessTokenNotAuthorizedException.cs:40]: To make the code simple.

**Review by vicancy — APPROVED** [review]

**Pull request merged**